include/Oscillator.h

```diff
@@ -153,6 +164,63 @@ class LMMS_EXPORT Oscillator
 		return m_userWave->userWaveSample( _sample );
 	}
 
+	struct wtSampleControl {
+		float frame;
+		f_cnt_t f1;
+		f_cnt_t f2;
+		int band;
+	};
+
+	inline wtSampleControl getWtSampleControl(const float sample)
+	{
+		wtSampleControl control;
+		control.frame = sample * OscillatorConstants::WAVETABLE_LENGTH;
+		control.f1 = static_cast<f_cnt_t>(control.frame) % OscillatorConstants::WAVETABLE_LENGTH;
+		if (control.f1 < 0)
+		{
+			control.f1 += OscillatorConstants::WAVETABLE_LENGTH;
+		}
+		control.f2 = control.f1 < OscillatorConstants::WAVETABLE_LENGTH - 1 ?
+					control.f1 + 1 :
+					0;
+		control.band = waveTableBandFromFreq(m_freq * m_detuning * Engine::mixer()->processingSampleRate());
+		return control;
+	}
+
+	inline sample_t wtSample(const sample_t table[OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT][OscillatorConstants::WAVETABLE_LENGTH], const float _sample)
+	{
+		wtSampleControl control = getWtSampleControl(_sample);
+		return linearInterpolate(table[control.band][control.f1],
+				table[control.band][control.f2], fraction(control.frame));
+	}
+
+	inline sample_t wtSample(const std::unique_ptr<OscillatorConstants::waveform_t>& table, const float sample)
+	{
+		wtSampleControl control = getWtSampleControl(sample);
+		return linearInterpolate((*table)[control.band][control.f1],
+				(*table)[control.band][control.f2], fraction(control.frame));
+	}
+
+	inline sample_t wtSample( sample_t **table, const float _sample)
+	{
+		wtSampleControl control = getWtSampleControl(_sample);
+		return linearInterpolate(table[control.band][control.f1],
+				table[control.band][control.f2], fraction(control.frame));
+	}
+
+	inline int  waveTableBandFromFreq(float freq)
+	{
+		int band = (69 + static_cast<int>(ceil(12.0f * log2f(freq / 440.0f)))) / OscillatorConstants::SEMITONES_PER_TABLE;
```

Comment:
Why?

Reply:
FWIR, `ceil` operates on double, while `ceilf` operates on floats. Here, the input is a float. Using double precision unnecessarily can make the program slower.

Reply:
`std::ceil` has a `float` overload while C-style `ceil` doesn't. The same applies to `round`, `floor`, etc.
I did some tests, and the behavior is not consistent: MSVC always uses the `float` variant, GCC and Clang uses `double` version when there are no statements like `using namespace std;` or `using std::ceil;`. Explicitly using functions in `std` namespace always uses optimal type, though.
FYI, the type of frequency may be changed to `double` in the future due to #2047.

Reply:
It's really a pain that we have `using namespace std;` in our headers. This makes it totally unclear which `ceil is meant`. If I understand you correctly, `std::ceil` should still be preferable to `ceilf`, because `std::ceil` is portable when we change to `double`?

Reply:
I will change both to `std::ceil` and `std::log2` if there are no objections. It seems to be the "cleanest C++ way".

As for a future change to double, this should not really matter, because this function only does quite coarse computation. IMO it should keep taking float even after "the double transition".